dkms.in

```diff
@@ -1040,22 +1107,52 @@ prepare_build()
             $"Check $build_dir for more information."
     done
 
-    if [[ -e "${kernel_config}" ]]; then
-        if grep -q 'CONFIG_CC_IS_CLANG=y' "${kernel_config}"; then
-            local cc=clang
-            if command -v "$cc" >/dev/null; then
-                export CC="$cc"
-                export KERNEL_CC="$cc"
+    # Detect CC and LD used by kernel
+    if ! grep -qE '(^|\s)CC=' <<< "${make_command}"; then
+        # No CC specified within MAKE in dkms.conf
+        if [[ -e "${kernel_config}" ]]; then
+            if grep -q '^CONFIG_CC_IS_GCC=y' "${kernel_config}"; then
+                # Since kernel 4.18
+                KERNEL_CC="$(detect_toolchain_from_kconfig GCC gcc)"
+            elif grep -q '^CONFIG_CC_IS_CLANG=y' "${kernel_config}"; then
+                # Since kernel 4.18
+                KERNEL_CC="$(detect_toolchain_from_kconfig CLANG clang)"
+            elif [[ -f "$kernel_source_dir/.kernelvariables" ]]; then
+                # Debian specific
+                KERNEL_CC="$(echo -e "show-%:\n\t@echo \$(\$*)\ninclude $kernel_source_dir/.kernelvariables" | make -f - show-CC)"
+            else
+                KERNEL_CC="$(detect_toolchain_from_cc_version_text)"
             fi
         fi
-
-        if grep -q 'CONFIG_LD_IS_LLD=y' "${kernel_config}"; then
-            local ld=ld.lld
-            if command -v "$ld" >/dev/null; then
-                export LD="$ld"
-                export KERNEL_LD="$ld"
+        if [[ -z "$KERNEL_CC" ]]; then
+            KERNEL_CC="$(detect_toolchain_from_vmlinux GCC gcc || \
+                         detect_toolchain_from_vmlinux clang clang)"
+        fi
+        if [[ "$KERNEL_CC" ]]; then
```

Comment:
Since `KERNEL_CC` and `KERNEL_LD` are local, so let's use `local KERNEL_CC...` in this function.

Both variables were exported with https://github.com/dell/dkms/pull/294, although searching through the [Debian patches/tooling](https://salsa.debian.org/kernel-team/linux) does not indicate why. So I suspect keeping them local is fine.